src/input/InputSDL.cpp

```diff
@@ -182,13 +191,41 @@ void InputSDLWorker::run()
       }
     }
 
+    m_mutex.lock();
+
+    unsigned long elapsed = (unsigned long)polltimer.elapsed();
+
     // make the poll time fixed to SDL_POLL_TIME
-    if (polltimer.elapsed() < SDL_POLL_TIME)
-      QThread::msleep(SDL_POLL_TIME - polltimer.elapsed());
+    if (elapsed < SDL_POLL_TIME)
+    {
+      unsigned long time = SDL_POLL_TIME - elapsed;
+      bool fg = QGuiApplication::focusWindow();
+      if (!fg)
+        time = ULONG_MAX;
+
+      m_waitCond.wait(&m_mutex, time);
+
+      if (!fg)
+      {
+        SDL_PumpEvents();
+        SDL_FlushEvents(SDL_APP_TERMINATING, SDL_JOYBUTTONUP);
+        SDL_FlushEvents(SDL_CONTROLLERAXISMOTION, SDL_LASTEVENT);
+      }
+    }
+
+    m_mutex.unlock();
   }
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////
+void InputSDLWorker::wake()
+{
+  m_mutex.lock();
+  m_waitCond.wakeAll();
+  m_mutex.unlock();
```

Comment:
Somewhat questionable use of condition variables. This is racy. It will wake up threads that are currently waiting, but not threads that are about to enter a wait call. It might actually be fine in this case, though - if you think it is, fine.